File: day6_go/day6b.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	input, _ := os.ReadFile("input.txt")
	lines := strings.Split(string(input), "\n")

	width := len(lines[0])
	height := len(lines)

	initalPosition := []int{0, 0}

	matrix := make([][]rune, len(lines))
	for i := range matrix {
		matrix[i] = make([]rune, len(lines[i]))
		for j := range lines[i] {
			matrix[i][j] = []rune(lines[i])[j]
			if matrix[i][j] == '^' {
				initalPosition[0] = i
				initalPosition[1] = j
			}
		}
	}

	// Let the guard walk one initial time
	{
		found := false
		direction := []int{-1, 0}
		position := []int{initalPosition[0], initalPosition[1]}

		for {
			if found {
				break
			}

			for {
				newPosition := []int{position[0] + direction[0], position[1] + direction[1]}
				if newPosition[0] < 0 || newPosition[1] < 0 || newPosition[0] >= width || newPosition[1] >= height {
					found = true
					break
				}

				if matrix[newPosition[0]][newPosition[1]] == '#' {
					break
				}

				matrix[newPosition[0]][newPosition[1]] = 'X'
				position = []int{newPosition[0], newPosition[1]}
			}
			direction = []int{direction[1], 0 - direction[0]}
		}
	}

	blocks := 0
	testMatrix := make([][]rune, height)
	for i := range testMatrix {
		testMatrix[i] = make([]rune, width)
	}

	for x := range width {
		for y := range height {
			if matrix[x][y] != 'X' {
				continue
			}

			for i := range testMatrix {
				copy(testMatrix[i], matrix[i])
			}

			found := false
			directionId := 0
			direction := []int{-1, 0}
			position := []int{initalPosition[0], initalPosition[1]}

			testMatrix[x][y] = '#'

			for {
				if found {
					break
				}

				for {
					newPosition := []int{position[0] + direction[0], position[1] + direction[1]}
					if newPosition[0] < 0 || newPosition[1] < 0 || newPosition[0] >= width || newPosition[1] >= height {
						found = true
						break
					}

					val := testMatrix[newPosition[0]][newPosition[1]]
					if val == '#' {
						break
					}

					// in a loop
					if val == rune(directionId) {
						blocks += 1
						found = true
						break
					}

					if val == '#' {
						break
					}

					testMatrix[newPosition[0]][newPosition[1]] = rune(directionId)

					position = []int{newPosition[0], newPosition[1]}
				}

				direction = []int{direction[1], 0 - direction[0]}
				directionId = (directionId + 1) % 4
			}
		}
	}

	fmt.Println(blocks)
}

File: day6/day6b.js
const fs = require('node:fs');
const input = fs.readFileSync(__dirname + '/input.txt', 'utf8').replace('\r', '').trim();

const matrix = input.split('\n').map((row) => row.split(''));

const width = matrix[0].length;
const height = matrix.length;

const inBound = (x, y) => x >= 0 && y >= 0 && x < width && y < height;

let initialPosition = [ 0, 0 ];
for (let x = 0; x < width; x++) {
    for (let y = 0; y < height; y++) {
        if (matrix[x][y] === '^') {
            matrix[x][y] = 0;
            initialPosition = [ x, y ];
        }
    }
}

// First fill the field with positions the player visits
(() => {
    let found = false;
    let direction = [ -1, 0 ];
    let position = initialPosition;
    while(found === false) {
        while(true) {
            const newX = position[0] + direction[0];
            const newY = position[1] + direction[1];
            if (!inBound(newX, newY)) {
                found = true;
                break
            }    
            if (matrix[newX][newY] === '#') {
                break;
            }
            matrix[newX][newY] = 'X';
            position = [newX, newY];        
        }
        direction = [ direction[1], 0-direction[0] ];
    }
})();

let blocks = 0;
for (let x = 0; x < width; x++) {
    for (let y = 0; y < height; y++) {
        // Check if the player ever visits that position
        if (matrix[x][y] !== 'X') {
            continue;
        }

        let found = false;

        let directionId = 0;
        let direction = [ -1, 0 ];
        let position = initialPosition;
        const testMatrix = matrix.map(line => [ ...line ]);

        testMatrix[x][y] = '#';

        while(found === false) {
            while(true) {
                const newX = position[0] + direction[0];
                const newY = position[1] + direction[1];
        
                if (!inBound(newX, newY)) {
                    found = true;
                    break
                }
        
                const val = testMatrix[newX][newY];
            
                // If we are traveling in the same direction we are
                // in a loop
                if (val === directionId) {
                    blocks++;
                    found = true;
                    break;
                }
        
                if (val === '#') {
                    break;
                }
        
                testMatrix[newX][newY] = directionId;
        
                position = [newX, newY];        
            }
        
            direction = [ direction[1], 0-direction[0] ];
            directionId = (directionId + 1) % 4;
        }
    }
}

console.log(blocks);
